contributors/design-proposals/key-set-as-merge-key.md

```diff
@@ -0,0 +1,336 @@
+# Key Set as Merge Key in Strategic Merge Patch
+
+## Abstract
+
+Support a key set as the merge key in Strategic Merge Patch.
+
+## Motivation
+
+*Background*: In Strategic Merge Patch, Merge Key is the key to distinguish the entries in the list of non-primitive types.
+It must always be present and unique to perform the merge on the list of non-primitive types,
+and will be preserved.
+
+The issue in practice is that sometimes we cannot find a key that always be present and unique.
+[#39188](https://github.com/kubernetes/kubernetes/issues/39188) is a typical example. We need to
+support a key set as Merge Key. A key set must contain at least 1 key.
+
+## Proposed Change
+
+### API Change
+
+We will update the struct tags for the API resources whose Merge Key is not guaranteed to be always
+present and unique. The keys will be seperated by "|", i.e. `patchMergeKey:"<key1>|<key2>|<key3>"`.
+
+E.g. `Ports` in `ServiceSpec`.
+```go
+type ServiceSpec struct {
+  // Change patchMergeKey from "port" to "name|port"
+  Ports []ServicePort `json:"ports,omitempty" patchStrategy:"merge" patchMergeKey:"name|port" protobuf:"bytes,1,rep,name=ports"`
+  ...
+}
+```
+
+All the impacted APIs are listed in section [Impacted APIs](#impacted-apis)
+
+### Open API
+
+Update Open API schema to have a string that contains multiple keys for `patchMergeKey`.
+The change should be trivial.
+
+### Strategic Merge Patch pkg
+
+Entries are considered as the same if and only if all the keys in the key set are identical.
+We allow keys to be missing in the key set as long as it is not empty.
```

Comment:
>If one or more of the mergeKeys is unspecified, it is interpreted as an empty value for purposes of uniquely identifying the object.

IMO we still want to distinguish unspecified and empty
```
- port: 12345
```
and
```
- name: ""
  port: 12345
```
The wording you give seems not support this?

Can you give me another wording?